tools/talis/digital_ocean.go

```diff
@@ -0,0 +1,397 @@
+	publicKey = strings.TrimSpace(publicKey)
+
+	// Pagination options
+	opt := &godo.ListOptions{PerPage: 200}
+
+	for {
+		keys, resp, err := client.Keys.List(ctx, opt)
+		if err != nil {
+			return godo.Key{}, fmt.Errorf("failed to list SSH keys: %w", err)
+		}
+
+		for _, key := range keys {
+			// Compare the trimmed public key string
+			if strings.TrimSpace(key.PublicKey) == publicKey {
+				return key, nil
+			}
+		}
+
+		// Break if we're at the last page
+		if resp.Links.IsLastPage() {
+			break
+		}
+		// Advance to next page
+		page, err := resp.Links.CurrentPage()
+		if err != nil {
+			return godo.Key{}, fmt.Errorf("unable to parse pagination: %w", err)
+		}
+		opt.Page = page + 1
+	}
+
+	return godo.Key{}, fmt.Errorf(
+		"ssh public key not found in DigitalOcean. Please upload it via the control panel or API before proceeding",
+	)
+}
+
+// CreateDroplets launches all droplets in parallel, waits for their IPs, and
+// returns the filled-out []Instance slice.
+func CreateDroplets(ctx context.Context, client *godo.Client, insts []Instance, key godo.Key) ([]Instance, error) {
+	total := len(insts)
+
+	type result struct {
+		inst         Instance
+		err          error
+		timeRequired time.Duration
+	}
+
+	insts, existing, err := filterExistingInstances(ctx, client, insts)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(existing) > 0 {
+		log.Println("Existing instances found, so they are not being created.")
+		for _, v := range existing {
+			log.Println("Skipping", v.Name, v.PublicIP, v.Tags)
+		}
+	}
+
+	results := make(chan result, total)
+	var wg sync.WaitGroup
+	wg.Add(total)
+
+	for _, v := range insts {
+		go func() {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(ctx, 7*time.Minute)
+			defer cancel()
+
+			req := &godo.DropletCreateRequest{
+				Name:   v.Name,
+				Region: v.Region,
+				Size:   v.Slug,
+				Image: godo.DropletCreateImage{
+					Slug: "ubuntu-22-04-x64",
+				},
+				SSHKeys: []godo.DropletCreateSSHKey{{ID: key.ID, Fingerprint: key.Fingerprint}},
+				Tags:    v.Tags,
+			}
+
+			start := time.Now()
+
+			log.Println("Creating droplet", v.Name, "in region", v.Region, start.Format(time.RFC3339))
+
+			d, _, err := client.Droplets.Create(ctx, req)
+			if err != nil {
+				results <- result{inst: v, err: fmt.Errorf("create %s: %w", v.Name, err)}
+				return
+			}
+
+			pubIP, privIP, err := waitForNetworkIP(ctx, client, d.ID)
+			if err != nil {
+				results <- result{inst: v, err: fmt.Errorf("public IP %s: %w", v.Name, err)}
+				return
+			}
+
+			v.PublicIP = pubIP
+			v.PrivateIP = privIP
+			results <- result{inst: v, err: nil, timeRequired: time.Since(start)}
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
+
+	var created []Instance
+	for res := range results {
+		if res.err != nil {
+			fmt.Printf("❌ %s failed after %v %v\n", res.inst.Name, res.timeRequired, res.err)
+		} else {
+			created = append(created, res.inst)
+			fmt.Printf("✅ %s is up (public=%s) in %v\n",
```

Comment:
[uber nit][not blocking] I think emojis can show up incorrectly in certain Terminal emulators so I would try to avoid using them in logs and print statements.

Reply:
they can use updated terminals or continue to hate fun imo